### PR TITLE
Add support for some AspNetCoreMetricsInstrumentationOptions

### PR DESCRIPTION
**Pull request by Temppus**
## Changes

Hi. I wanted to include standard http metrics via `AddOpenTelemetryMetrics` and it works great. But I am missing option to add additional tags to current metrics recorded. In my case I wanted to enrich metrics with tag of current tenant.
I saw there was no support for it yet.

**Notes:**

- new option class specific for metrics **AspNetCoreMetricsInstrumentationOptions** was introduced
- added possibility for filtering/enrichment

**Tests:**

- Added test for both filtering and enrichment

I would like to here your feedback, thanks.

TODO:
* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [x] Design discussion issue #
* [x] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreMetricsInstrumentationOptions.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreMetricsInstrumentationOptions.cs:43]: * nit: We could probably just call this `Enrich`
* Let's change the signature to be: `Func<HttpContext, System.Diagnostics.TagList>`. The nice thing about `TagList` is it is a struct so this can be done without an allocation for a certain number of tags (8 at the moment, IIRC).

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:57]: We should wrap the calls into user code with try/catch. Here's a couple examples you can follow:

* https://github.com/open-telemetry/opentelemetry-dotnet/blob/39283c7327435b1d3ca9702387363030226b45a6/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L174-L190

* https://github.com/open-telemetry/opentelemetry-dotnet/blob/39283c7327435b1d3ca9702387363030226b45a6/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L223-L230

File: src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreMetricsInstrumentationOptions.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreMetricsInstrumentationOptions.cs:42]: Might be nice if we mentioned the timing here. Something like this...

```xml
        /// <remarks>
        /// Note: Enrich is fired after the request has completed.
        /// <para><see cref="HttpContext"/>: the HttpContext object.</para>
        /// </remarks>
```

Trying to give users a hint that Request & Response are both available when the callback fires 🤷

File: src/OpenTelemetry.Instrumentation.AspNetCore/MeterProviderBuilderExtensions.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNetCore/MeterProviderBuilderExtensions.cs:75]: nit: Do me a favor and remove this TODO the ConfigureBuilder call satisfies IDeferredMeterProviderBuilder.

File: src/OpenTelemetry.Instrumentation.AspNetCore/.publicApi/net7.0/PublicAPI.Unshipped.txt
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNetCore/.publicApi/net7.0/PublicAPI.Unshipped.txt:15]: FYI in order to get the CI passing you'll need to update this file for the other targets as well (net6.0, netcoreapp3.1, netstandard2.0, & netstandard2.1).

**Comment by CodeBlanch** [comment]
@Temppus Please add a note in the project CHANGELOG about the change.

File: src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreMetricsInstrumentationOptions.cs
Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreMetricsInstrumentationOptions.cs:43]: Actually on the signature, if you want to achieve max perf do something like this:

```csharp
public class AspNetCoreMetricsInstrumentationOptions
{
        public delegate void AspNetCoreMetricEnrichmentFunc(HttpContext context, out TagList tags);

        public AspNetCoreMetricEnrichmentFunc Enrich { get; set; }
}
```

* `TagList` ~avoids allocations.
* The `out` usage avoids copying a potentially very big `struct`.

PS: In that snippet I nested the delegate in the options class, I think that is probably fine to do for this.

Comment (Temppus) [reply src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreMetricsInstrumentationOptions.cs]: Reworked as suggested

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs
Comment (Temppus) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs]: Added as suggested

File: src/OpenTelemetry.Instrumentation.AspNetCore/MeterProviderBuilderExtensions.cs
Comment (Temppus) [reply src/OpenTelemetry.Instrumentation.AspNetCore/MeterProviderBuilderExtensions.cs]: Removed

File: src/OpenTelemetry.Instrumentation.AspNetCore/.publicApi/net7.0/PublicAPI.Unshipped.txt
Comment (Temppus) [reply src/OpenTelemetry.Instrumentation.AspNetCore/.publicApi/net7.0/PublicAPI.Unshipped.txt:15]: Updated to all targets

**Comment by Temppus** [comment]
@CodeBlanch Changelog updated

File: src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreMetricsInstrumentationOptions.cs
Comment (Temppus) [reply src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreMetricsInstrumentationOptions.cs]: Added

**Review by CodeBlanch — APPROVED** [review]
LGTM. We should probably update README with some details but that can be done as a follow-up. The current content is at least close 😄

@Temppus Just FYI I'm going to let this sit for a while to see if any other maintainers comment. Lots of people out on vacation.

File: src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md:23]: nits:

* Extra space at the beginning.

* "fot" 🤣

**Comment by alanwest** [comment]
I’m becoming suspicious of the Filter and maybe also Enrich callbacks…

For traces I think Filter provides a poor approximation of a sampler. For example, say there's an endpoint that makes a call using HttpClient. If Filter is used to omit spans to this endpoint, often the user would expect the downstream HttpClient span to also be omitted. However, depending on the sampler in use, Filter cannot prevent the HttpClient span from being exported. Makes it seem like a custom sampler would be preferable.

That said, Filter seems more appealing for metrics given metrics have no notion of a sampler. Though, I wonder if using `Sdk.SuppressInstrumentation` would provide a better solution for users wanting to filter certain metrics. `Sdk.SuppressInstrumentation` could also have the benefit of suppressing the downstream HttpClient metric. @dnelson-relativity shared a [middleware solution](https://github.com/dnelson-relativity/asp-net-core-route-suppression/blob/main/AspNetCoreRouteSuppression/SpanSuppressionMiddleware.cs) using `Sdk.SuppressInstrumentation`.

I'm a little less suspicious of the value of Enrich, however, I suspect there are alternate solutions for many of the use cases for which Enrich is used.

> In my case I wanted to enrich metrics with tag of current tenant.

There have been discussions in the spec to support this kind of scenario with either the View API and baggage (see #2619) or by introducing the concept of a measurement processor (much like a span processor but is invoked when a measurement is recorded).

I'll put this PR on tomorrow's SIG meeting agenda for discussion (11 am Pacific).

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:106]: @alanwest @utpilla One question. The current design we call `Enrich` which returns tags (using `out` parameter). The returned tags are added to the list curated by the instrumentation. This is subtly different from the `Activity` enrich callbacks where we pass the `Activity` allowing for full modification of tags added by the instrumentation. Perhaps instead of `out` we should pass the `tags` above as a `ref` to the callback which would allow for addition & modification akin to tracing?

**Comment by Temppus** [comment]
> I'll put this PR on tomorrow's SIG meeting agenda for discussion (11 am Pacific).

Hi, what was the outcome of it ?

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs
Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:106]: > Perhaps instead of out we should pass the tags above as a ref to the callback which would allow for addition & modification akin to tracing?

I think that's a good idea.

File: src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md
Comment (Temppus) [reply src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md]: fixed

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs
Comment (Temppus) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs]: Reworked as ref

**Comment by Temppus** [comment]
Can you rerun the CI again ? Thanks.

**Comment by Temppus** [comment]
> Can you rerun the CI again ? Thanks.
one more time pls  :)

**Comment by Temppus** [comment]
> Can you rerun the CI again ? Thanks.

Hopefully last time ...

**Comment by Temppus** [comment]
Is [TestRecoveryAfterFailedExport ](https://github.com/open-telemetry/opentelemetry-dotnet/actions/runs/3624179701/jobs/6111254066) flaky ? It seems totally unrelated to my changes.

**Comment by utpilla** [comment]
> Is [TestRecoveryAfterFailedExport ](https://github.com/open-telemetry/opentelemetry-dotnet/actions/runs/3624179701/jobs/6111254066) flaky ? It seems totally unrelated to my changes.

Yes, the Code Coverage check is flaky.

**Comment by Temppus** [comment]
> Yes, the Code Coverage check is flaky.

So there is a chance that when CI will be rerun it it will pass ?

**Comment by Kielek** [comment]
> > Yes, the Code Coverage check is flaky.
> 
> So there is a chance that when CI will be rerun it it will pass ?

Don't bother about Code coverage action if there is not issue related to your code. It is not mandatory to pass before merge (as long as it is not stable). Other steps are passing.

**Comment by Temppus** [comment]
@CodeBlanch is there anything that should be changed to merge this ?

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:60]: nit: the log does not clearly indicates that the measurement is filtered out.

File: src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md:23]:
```suggestion
   configuration properties for filtering and tag enrichment of `http.server.duration` metric.
```

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs
Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:60]: I put a fix on #3982 let me know if that works!

File: src/OpenTelemetry.Instrumentation.AspNetCore/.publicApi/netcoreapp3.1/PublicAPI.Unshipped.txt
Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.AspNetCore/.publicApi/netcoreapp3.1/PublicAPI.Unshipped.txt:11]: Fixed on #3982

**Pull request merged**
